torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1920,14 +1987,76 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
         if not self._suppress_guards_tls():
             stack = ''.join(traceback.format_list(traceback.extract_stack()[:-2]))
             if concrete_val is sympy.true:
-                self.guards.append(ShapeGuard(expr, stack))
+                guard = ShapeGuard(expr, stack)
             elif concrete_val is sympy.false:
-                self.guards.append(ShapeGuard(sympy.Not(expr), stack))
+                guard = ShapeGuard(sympy.Not(expr), stack)
             else:
-                self.guards.append(
-                    ShapeGuard(sympy.Eq(expr, concrete_val), stack))  # type: ignore[arg-type]
+                guard = ShapeGuard(sympy.Eq(expr, concrete_val), stack)  # type: ignore[arg-type]
+
+            self.guards.append(guard)
+            self.refine_ranges(guard)
+
         return concrete_val
 
+    def refine_ranges(self, guard: ShapeGuard) -> None:
+        def simplify(expr: sympy.Expr) -> sympy.Expr:
+            """
+            Simplification specialized for range refinement.
+            """
+            return self.maybe_simplify_floordiv(self.simplify(expr))
+
+        # First, try to simplify the guard.
+        # Since we do not guarantee that the function is monotonic,
+        # we set a maximum number of iterations.
+        max_iterations = 10
+        it = 0
+        expr = guard.expr
+
+        while True:
+            it += 1
+            previous = expr
+            expr = simplify(previous)
+            if expr == previous or it >= max_iterations:
+                break
+
+        # Filter the guards that:
+        #   1. are relational operations
+        #   2. have a symbol as the left-hand side
```

Comment:
Can the symbol show up on rhs or is that impossible?

Reply:
I believe they can. So, here, we are being restrictive: `maybe_simplify_floordiv` tries to move stuff to the right so that we are left with a symbol in the left.

Reply:
We could, however, invert the expression (e.g. `a < b` into `b > a`) and re-run the value refinement procedure again. How does this sound?

Reply:
I have implemented this idea. Let me know what you think.